src/AspNet.Security.OAuth.Alipay/AlipayAuthenticationHandler.cs

```diff
@@ -0,0 +1,231 @@
+            };
+            sortedParams.Add("sign", GetRSA2Signature(sortedParams));
+
+            string address = QueryHelpers.AddQueryString(Options.UserInformationEndpoint, sortedParams);
+
+            using var response = await Backchannel.GetAsync(address);
+
+            if (!response.IsSuccessStatusCode)
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                                "returned a {Status} response with the following payload: {Headers} {Body}.",
+                                /* Status: */ response.StatusCode,
+                                /* Headers: */ response.Headers.ToString(),
+                                /* Body: */ await response.Content.ReadAsStringAsync());
+
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+
+            using var payload = await ParseJsonAsync(response, "alipay_user_info_share_response");
+            string statusCode = payload.RootElement.GetString("code");
+
+            if (statusCode != "10000")/* if code==10000 then success else failed. See: https://docs.open.alipay.com/common/105806 */
+            {
+                Logger.LogError("An error occurred while retrieving the user profile: the remote server " +
+                            "returned a {Status} response with the following message: {Message}.",
+                            /* Status: */ statusCode,
+                            /* Message: */ payload.RootElement.GetString("msg"));
+
+                throw new HttpRequestException("An error occurred while retrieving user information.");
+            }
+
+            identity.AddClaim(new Claim(ClaimTypes.NameIdentifier, payload.RootElement.GetString("user_id"), ClaimValueTypes.String, Options.ClaimsIssuer));
+
+            var principal = new ClaimsPrincipal(identity);
+            var context = new OAuthCreatingTicketContext(principal, properties, Context, Scheme, Options, Backchannel, tokens, payload.RootElement);
+
+            context.RunClaimActions();
+
+            await Options.Events.CreatingTicket(context);
+            return new AuthenticationTicket(context.Principal, context.Properties, Scheme.Name);
+        }
+
+        /// <summary>
+        /// Gets an HTTP payload as an asynchronous operation.
+        /// </summary>
+        /// <param name="message">The message.</param>
+        /// <param name="elementName">Name of the element.</param>
+        /// <returns>Task&lt;JsonDocument&gt;.</returns>
+        private async Task<JsonDocument> ParseJsonAsync(HttpResponseMessage message, string elementName)
+        {
+            var messageStream = await message.Content.ReadAsStreamAsync();
+
+            var document = await JsonSerializer.DeserializeAsync<JsonElement>(messageStream);
+            var elementContent = document.GetProperty(elementName).GetRawText();
```

Comment:
I know `System.Text.Json`'s limited API surface sometimes requires doing weird things, but in this particular case, I still don't understand why you need to export the RAW JSON of your `elementName` node to re-parse it. Why not directly returning `elementContent`?

Reply:
public static OAuthTokenResponse Success(JsonDocument response);

JsonElement must be convert to JsonDocument.

Reply:
Understood, but it's only required in `ExchangeCodeAsync`, not in `CreateTicketAsync` where you use `payload.RootElement`. Let's get rid of this private method and only do the `GetRawText()` trick for `ExchangeCodeAsync`.

Reply:
It's my mistake. 
how about: 
            var elementContent = document.GetString(elementName);
            return JsonDocument.Parse(elementContent);

Reply:
`GetString` will only work if the `elementName` node is a string. If it's a JSON object or array, it will fail.

Reply:
> ValueKind = Object : "{
  "alipay_system_oauth_token_response": {
    "user_id": "my-id",
    "access_token": "secret-access-token",
    "expires_in": "300",
    "refresh_token": "secret-refresh-token",
    "re_expires_in": "300"
  },
  "sign": "sign"
}"

var elementContent = document.GetString(elementName);
Return:

> {
    "user_id": "my-id",
    "access_token": "secret-access-token",
    "expires_in": "300",
    "refresh_token": "secret-refresh-token",
    "re_expires_in": "300"
  }

Seem to work.

Reply:
Oh yeah, that's the ASP.NET Core extension, not `JsonElement.GetString()`: https://github.com/aspnet/AspNetCore/blob/master/src/Security/Authentication/Core/src/JsonDocumentAuthExtensions.cs#L10

I'd keep `GetRawString()` to reduce potential confusion here.